pkg/controller/telemetry/map_metric.go

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/cilium/ebpf"
+)
+
+const (
+	mapMetricFlushInterval = 20 * time.Second
+)
+
+type MapMetricController struct {
+	mapMetricCache map[mapMetricLabels]*mapUsageInfo
+}
+
+type mapUsageMetric struct {
+	mapId      uint32
+	mapName    string
+	mapType    string
+	maxEntries uint32
+	keySize    uint32
+	valueSize  uint32
+	memLock    uint64
+	entryCount uint32
+}
+
+type mapUsageInfo struct {
+	memLock    uint64
+	entryCount uint32
+	maxEntries uint32
+}
+
+type mapMetricLabels struct {
+	mapId        string
+	mapName      string
+	mapType      string
+	podName      string
+	podNamespace string
+}
+
+func NewMapMetric() *MapMetricController {
+	return &MapMetricController{
+		mapMetricCache: map[mapMetricLabels]*mapUsageInfo{},
+	}
+}
+
+func (m *MapMetricController) Run(ctx context.Context) {
+	if m == nil {
+		return
+	}
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// Metrics updated every 5 seconds
+				time.Sleep(mapMetricFlushInterval)
+				m.updatePrometheusMetric()
+			}
+		}
+	}()
+}
+
+func (m *MapMetricController) buildMapMetric(data *mapUsageMetric) mapMetricLabels {
+	labels := mapMetricLabels{}
+	// Get the actual pod namespace from environment variable "POD_NAMESPACE"
+	podNamespace := os.Getenv("POD_NAMESPACE")
+	if podNamespace == "" {
+		podNamespace = "unknown-pod-namespace"
+	}
+	podName := os.Getenv("POD_NAME")
+	if podName == "" {
+		podName = "unknown-pod-name"
+	}
+
+	labels.podName = podName
+	labels.podNamespace = podNamespace
+	labels.mapId = fmt.Sprintf("%d", data.mapId)
+	labels.mapName = data.mapName
+	labels.mapType = data.mapType
+	return labels
+}
+
+func (m *MapMetricController) updatePrometheusMetric() {
+	var startID ebpf.MapID
+	count := 0
+	currentMapIDs := make(map[string]bool)
+	var commonLabels map[string]string
+
+	for {
+		mapID, mapInfo, info, err := getNextMapInfo(startID)
```

Comment:
IIUC, this is iterating all the maps within the node, but how could we know if a map belong to kmesh?